### PR TITLE
StatsD connections should control the endpoint

### PR DESCRIPTION
**Pull request by trevnorris**
The main change here is for each connection to control its own reconnection logic. That'll make it easier to abstract the implementation out for a virtual class.

### CODE REVIEW

File: agents/statsd/src/statsd_agent.cc [hunk 1]
```diff
@@ -697,9 +697,10 @@ int StatsDAgent::config_handles() {
     ASSERT_EQ(0, retry_timer_.stop());
     if (endpoint_->protocol() == "tcp") {
       setup_tcp();
-    } else {
-      ASSERT_STR_EQ(endpoint_->protocol().c_str(), "udp");
+    } else if (endpoint_->protocol() == "udp") {
       setup_udp();
+    } else {
+      abort();  // Unreachable
```
Comment (santigimeno) [inline agents/statsd/src/statsd_agent.cc]: Why this change? IIUC the current code behaves the same but the assertion message will give print the wrong protocol

File: agents/statsd/src/statsd_agent.cc [hunk 2]
```diff
@@ -997,9 +998,10 @@ void StatsDAgent::retry_timer_cb_(nsuv::ns_timer*, WeakStatsDAgent agent_wp) {
   ASSERT_NOT_NULL(agent->endpoint_.get());
   if (agent->endpoint_->protocol() == "tcp") {
     agent->setup_tcp();
-  } else {
-    ASSERT_STR_EQ(agent->endpoint_->protocol().c_str(), "udp");
+  } else if (endpoint_->protocol() == "udp") {
     agent->setup_udp();
+  } else {
+    abort();  // Unreachable
```
Comment (santigimeno) [inline agents/statsd/src/statsd_agent.cc]: Ditto

File: agents/statsd/src/statsd_agent.cc [hunk 3]
```diff
@@ -39,21 +39,28 @@ inline void DebugJSON(const char* str, const json& msg) {
...
+  ASSERT_EQ(0, retry_timer_->init(loop));
   ASSERT_EQ(0, addr_str_lock_.init());
   write_req_.set_data(nullptr);
 }
 
 StatsDTcp::~StatsDTcp() {
   addr_str_lock_.destroy();
+  // Make sure close_and_delete() is always called before the destructor is
+  // called.
+  ASSERT_NULL(retry_timer_);
```
Comment (santigimeno) [inline agents/statsd/src/statsd_agent.cc:54]: Should we also check for `tcp_` to be NULL?

File: agents/statsd/src/statsd_agent.cc [hunk 4]
```diff
@@ -118,6 +106,26 @@ int StatsDTcp::write(const string_vector& messages) {
   return r;
 }
 
+void StatsDTcp::setup(StatsDEndpoint* endpoint) {
+  ASSERT_NOT_NULL(endpoint);
+
+  int r = retry_timer_->stop();
+  if (r) {
+    // Nothing to do.
+  }
```
Comment (santigimeno) [inline agents/statsd/src/statsd_agent.cc]: We should add a `USE()` macro, or maybe just use the one from node (though it's in internals)

File: agents/statsd/src/statsd_agent.cc [hunk 5]
```diff
@@ -118,6 +106,26 @@ int StatsDTcp::write(const string_vector& messages) {
...
+  if (tcp_ != nullptr) {
+    tcp_->close_and_delete();
+    tcp_ = nullptr;
```
Comment (santigimeno) [inline agents/statsd/src/statsd_agent.cc]: no need to set tcp_ to null here

File: agents/statsd/src/statsd_agent.h [hunk 6]
```diff
@@ -34,8 +36,13 @@ static const nlohmann::json default_agent_config = {
   { "pauseMetrics", false }
 };
 
-class StatsDTcp {
+using WeakStatsDTcp = std::weak_ptr<StatsDTcp>;
+
+class StatsDTcp : public std::enable_shared_from_this<StatsDTcp> {
```
Comment (santigimeno) [inline agents/statsd/src/statsd_agent.h]: IIUC the [documentation](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this), If you're inheriting from `std::enable_shared_from_this` the `StatsDTcp` instance should be stored in a `shared_ptr`.

File: agents/statsd/src/statsd_agent.cc [hunk 7]
```diff
@@ -280,6 +325,25 @@ int StatsDUdp::write(const string_vector& messages) {
...
+
+  int r = retry_timer_->stop();
+  if (r) {
+    // Nothing to do.
+  }
+  if (udp_ != nullptr) {
+    udp_->close_and_delete();
+    udp_ = nullptr;
+  }
+  udp_ = new nsuv::ns_udp();
```
Comment (santigimeno) [inline agents/statsd/src/statsd_agent.cc]:
```suggestion
  if (udp_ != nullptr) {
    udp_->close_and_delete();
  }
  udp_ = new nsuv::ns_udp();
```

**Review by santigimeno — CHANGES_REQUESTED** [review]
Mostly LGTM. Left some comments

File: agents/statsd/src/statsd_agent.h (hunk 6 above)
Comment (trevnorris) [reply agents/statsd/src/statsd_agent.h]: Looking over it more carefully, I realize now that it's not necessary to inherit from `std::enable_shared_from_this` if the life of the timer is controlled more carefully. Going to do the rewrite.

File: agents/statsd/src/statsd_agent.cc (hunk 1 above)
Comment (trevnorris) [reply agents/statsd/src/statsd_agent.cc]: The next PR will abstract away `tcp_` and `udp_` to a single virtual class stored by the `StatsDAgent`. Do we want to put a `Debug()` message before this so it prints the endpoint information?

File: agents/statsd/src/statsd_agent.cc (hunk 2 above)
Comment (trevnorris) [reply agents/statsd/src/statsd_agent.cc]: Same here about adding the `Debug()` message.

File: agents/statsd/src/statsd_agent.cc (hunk 3 above)
Comment (trevnorris) [reply agents/statsd/src/statsd_agent.cc:54]: Sure.

File: agents/statsd/src/statsd_agent.cc (hunk 4 above)
Comment (trevnorris) [reply agents/statsd/src/statsd_agent.cc]: I forgot that was a thing. I'll use that.

File: agents/statsd/src/statsd_agent.cc (hunk 5 above)
Comment (trevnorris) [reply agents/statsd/src/statsd_agent.cc]: will do

File: agents/statsd/src/statsd_agent.cc (hunk 7 above)
Comment (trevnorris) [reply agents/statsd/src/statsd_agent.cc]: will do

File: agents/statsd/src/statsd_agent.cc [hunk 8]
```diff
@@ -161,18 +164,18 @@ void StatsDTcp::connect_cb_(nsuv::ns_connect<nsuv::ns_tcp>* req,
...
   }
-  tcp->connect_();
+
+  Debug("Error connecting to all addresses. Retry in 3 seconds...\n");
+  tcp->status(ConnectionError);
+  // Reattempt all IP addresses after 3 seconds.
+  tcp->addr_index_ = 0;
+  tcp->prep_retry_timer_();
+  ASSERT_EQ(0, tcp->retry_timer_->start(
+        StatsDTcp::retry_timer_cb_, 3000, 0, tcp));
```
Comment (santigimeno) [inline agents/statsd/src/statsd_agent.cc]: Nit: can this logic refactor into a helper function ?

**Review by santigimeno — APPROVED** [review]

File: agents/statsd/src/statsd_agent.cc (hunk 8 above)
Comment (trevnorris) [reply agents/statsd/src/statsd_agent.cc]: will do
